### PR TITLE
SNOW-1023211: Support resample with rule offset `W`, `ME`, `YE` with `closed = left`

### PR DESCRIPTION
**Pull request by sfc-gh-nkrishna**
SNOW-1023211

This PR adds support for `resample` by rule offset `W`, `ME`, `YE` with parameter `closed = left`. 

Note: Snowpark pandas `resample` implementation uses the Snowflake SQL function `TIME_SLICE` which assumes the left hand side of the given interval is closed, meaning support using `closed = right` parameter is non-trivial.

### CODE REVIEW

File: tests/integ/modin/resample/test_resample.py [hunk 1]
```diff
@@ -41,7 +41,7 @@ def test_resample_with_varying_freq_and_interval(freq, interval, agg_func):
             {"A": np.random.randn(15)},
             index=native_pd.date_range("2020-01-01", periods=15, freq=f"1{freq}"),
         ),
-        lambda df: getattr(df.resample(rule=rule), agg_func)(),
+        lambda df: getattr(df.resample(rule=rule, closed="left"), agg_func)(),
```
Comment (sfc-gh-nkrishna) [inline tests/integ/modin/resample/test_resample.py:44]: These tests are all parametrized by the `freq` parameter which uses `IMPLEMENTED_DATEOFFSET_STRINGS` from `resample_utils.py`, which now includes the offsets supported in this PR. 

There are other tests in this file that specifically use freq `s`, `min`, `D`, `H` which tests that `closed = None` (i.e. the default closed value) still works for those frequencies.

File: docs/source/modin/supported/dataframe_supported.rst [hunk 2]
```diff
@@ -347,10 +348,11 @@ Methods
...
-| ``resample``                | P                               | ``axis``, ``closed``, ``label``, | Only DatetimeIndex is supported and its ``freq``   |
-|                             |                                 | ``convention``, ``kind``, ``on`` | will be lost. Only ``rule`` frequencies 's', 'min',|
-|                             |                                 | , ``level``, ``origin``,         | 'h', and 'D' are supported.                        |
-|                             |                                 | , ``offset``, ``group_keys``     |                                                    |
+| ``resample``                | P                               | ``axis``, ``label``,             | Only DatetimeIndex is supported and its ``freq``   |
+|                             |                                 | ``convention``, ``kind``, ``on`` | will be lost. ``rule`` frequencies 's', 'min',     |
+|                             |                                 | , ``level``, ``origin``,         | 'h', and 'D' are supported. ``rule`` frequency 'W' |
```
Comment (sfc-gh-helmeleegy) [inline docs/source/modin/supported/dataframe_supported.rst:353]:
```suggestion
|                             |                                 | , ``level``, ``origin``,         | 'h', and 'D' are supported. ``rule`` frequencies   |
```

File: docs/source/modin/supported/dataframe_supported.rst [hunk 3]
```diff
@@ -347,10 +348,11 @@ Methods
...
+|                             |                                 | , ``offset``, ``group_keys``     | , 'ME', and 'YE' are supported with                |
```
Comment (sfc-gh-helmeleegy) [inline docs/source/modin/supported/dataframe_supported.rst:354]:
```suggestion
|                             |                                 | , ``offset``, ``group_keys``     | 'W', 'ME', and 'YE' are supported with             |
```

File: docs/source/modin/supported/series_supported.rst [hunk 4]
```diff
@@ -340,10 +341,11 @@ Methods
...
-| ``resample``                | P                               | ``axis``, ``closed``, ``label``, | Only DatetimeIndex is supported and its ``freq``   |
-|                             |                                 | ``convention``, ``kind``, ``on`` | will be lost. Only ``rule`` frequencies 's', 'min',|
-|                             |                                 | , ``level``, ``origin``,         | 'h', and 'D' are supported.                        |
-|                             |                                 | , ``offset``, ``group_keys``     |                                                    |
+| ``resample``                | P                               | ``axis``, ``label``,             | Only DatetimeIndex is supported and its ``freq``   |
+|                             |                                 | ``convention``, ``kind``, ``on`` | will be lost. ``rule`` frequencies 's', 'min',     |
+|                             |                                 | , ``level``, ``origin``,         | 'h', and 'D' are supported. ``rule`` frequency 'W' |
```
Comment (sfc-gh-helmeleegy) [inline docs/source/modin/supported/series_supported.rst:346]:
```suggestion
|                             |                                 | , ``level``, ``origin``,         | 'h', and 'D' are supported. ``rule`` frequencies   |
```

File: docs/source/modin/supported/series_supported.rst [hunk 5]
```diff
@@ -340,10 +341,11 @@ Methods
...
+|                             |                                 | , ``offset``, ``group_keys``     | , 'ME', and 'YE' are supported with                |
```
Comment (sfc-gh-helmeleegy) [inline docs/source/modin/supported/series_supported.rst:347]:
```suggestion
|                             |                                 | , ``offset``, ``group_keys``     | 'W', 'ME', and 'YE' are supported with             |
```

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py [hunk 6]
```diff
@@ -11987,14 +11996,41 @@ def resample(
...
+                    min_max_index_column_quoted_identifier[0]
+                ),
+                date_trunc(slice_unit, max_(snowflake_index_column_identifier)).as_(
+                    min_max_index_column_quoted_identifier[1]
+                ),
+            ).collect()[0]
+        else:
+            assert slice_unit in RULE_WEEK_TO_YEAR
+            start_date, end_date = frame.ordered_dataframe.agg(
+                last_day(
```
Comment (sfc-gh-nkumar) [inline src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:12011]: maybe add a comment here explaining the logic?

File: src/snowflake/snowpark/modin/plugin/_internal/resample_utils.py [hunk 7]
```diff
@@ -82,9 +84,21 @@
...
+    "MS",
+    "ME",
+    "QS",
+    "QE",
+    "YS",
+    "YE",
+]
 
-IMPLEMENTED_DATEOFFSET_STRINGS = ["min", "s", "h", "D"]
+IMPLEMENTED_DATEOFFSET_STRINGS = ["s", "min", "h", "D", "W", "ME", "YE"]
```
[the remark this answers is not in the document]
Comment (sfc-gh-nkrishna) [reply src/snowflake/snowpark/modin/plugin/_internal/resample_utils.py:101]: We use the difference between the two in `test_resample_negative.py::test_resample_not_yet_implemented_freq` to test that other DateOffsets that the Snowpark function `time_slice` supports but we do not in Snowpark pandas throw the correct `NotImplementedError`.

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py (hunk 6 above)
Comment (sfc-gh-nkrishna) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:12015]: Added a comment.

**Review by sfc-gh-helmeleegy — APPROVED** [review]
LGTM.

**Review by sfc-gh-nkumar — APPROVED** [review]

**Pull request merged**